New tekton.dev Experience (PR 2 of 2: Implementation)

michaelawyu opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

~~Note: Netlify preview will **NOT** work properly as the instance has not been set up with proper credentials. [An alternate preview is available here (via Firebase)](https://tekton-website-258208.firebaseapp.com/).~~

Update: The Netlify preview is now working ([Link](https://deploy-preview-24--tekton-dev.netlify.com/)).

This PR implements the new tekton.dev website on top of the baseline
templates applied in [an earlier PR](https://github.com/tektoncd/website/pull/14).

This PR, as discussed with the team, is the second (and the last) of two PRs
that implements the new tekton.dev experience. It is still **in progress**,
pending the resolution of a few procedural issues. See the progress tracker below
for more information:

## Tech

- [x] New tekton.dev experience with [Hugo](https://gohugo.io) and [Google Hugo Docsy Theme](https://github.com/google/docsy)
- [x] Support for multiple documentation versions
- [x] Support for content synchronization
- [x] Support for content curation
- [x] CI/CD workflow with Tekton for website generation
- [x] Netlify support
- [x] Google Analytics integration
- [x] Google Custom Search integration
- [x] Misc.

## Design

- [x] Cover
- [x] Asciinema cast
- [x] Partner logo showcase
- [x] Manifesto
- [x] Interative tutorial showcase (tekton.dev/try)
- [x] Version switcher
- [x] Cookie usage notice for Google Analytics integration
- [x] Misc.

## Contents

- [x] Formatting contents in other sources of truth
- [x] Interactive tutorials for launch

## Legal

- [x] Permission to use selected background image
- [N/A] Permission to use partner logos
- [x] Permission to integrate Google Analytics

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [N/A] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [Done] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/website/blob/master/CONTRIBUTING.md)
for more details._

michaelawyu commented:
Note: Netlify preview **will NOT** work properly as the instance has not been set up with proper credentials.

carlos-logro commented:
Website looks to be heading in the right direction. Great work Chen!😸
I haven't gone around to review all the files but I did check the latest version and noticed a couple nits that should be worked on before releasing.

## 1) In mobile view or smaller width size the search bar disappears. Is this intended?
<img width="471" alt="Screen Shot 2020-01-23 at 3 31 04 PM" src="https://user-images.githubusercontent.com/49996607/73021469-a99ad580-3df5-11ea-8630-90f6fbab3a96.png">

## 2) In mobile view or smaller width size this should be centralized
<img width="524" alt="Screen Shot 2020-01-23 at 3 31 13 PM" src="https://user-images.githubusercontent.com/49996607/73021471-a99ad580-3df5-11ea-842b-e0623f268643.png">

## 3) In mobile view or smaller width I'd suggest adding some margin or padding between each company
<img width="485" alt="Screen Shot 2020-01-23 at 3 31 24 PM" src="https://user-images.githubusercontent.com/49996607/73021472-a99ad580-3df5-11ea-88bf-26d712f3ba6d.png">

## 4)`Try page` with its current design and styles seems to not belong in the website and/or is inconsistent with the other pages. For example it should at least have the same header/nav bar that all the other pages have. 

<img width="1679" alt="Screen Shot 2020-01-23 at 3 41 37 PM" src="https://user-images.githubusercontent.com/49996607/73022090-e6b39780-3df6-11ea-8885-bb120c835cdf.png">

### When viewing a tutorial on mobile devices or smaller screen width, it appears to be hard to navigate through 
<img width="462" alt="Screen Shot 2020-01-23 at 3 32 06 PM" src="https://user-images.githubusercontent.com/49996607/73021473-a99ad580-3df5-11ea-8f40-719ce76740d1.png">

### I feel the terminal in the background should not be visible until the user clicks 'Start Scenerio`
<img width="1668" alt="Screen Shot 2020-01-23 at 3 32 33 PM" src="https://user-images.githubusercontent.com/49996607/73021474-aa336c00-3df5-11ea-8815-eadc1ddc6494.png">

bobcatfish commented on helper/helper.py:
```diff
@@ -0,0 +1,124 @@
+# This script helps synchronize contents from their respective sources of
+# truth (usually GitHub repositories of each Tekton
+# components, such as tektoncd/pipelines) to tektoncd/website.
```
sounds like the "helper" is intended to sync the docs contents - could we name it something a bit more descriptive, e.g. `sync` or `docs-update` or `fetcher`?

michaelawyu replied:
Got it.

bobcatfish commented on k8s/README.md:
```diff
@@ -0,0 +1,7 @@
+- Clone the website from the tektoncd/website repository
+- Synchronizes contents from specified Tekton repositories
+- Build the website with Hugo
+- Deploys thew website via Firebase
```
since we're using netlify now, i'm thinking maybe we don't need these?

michaelawyu replied:
I'll remove this part.

bobcatfish commented on asciinema/README.md:
````diff
@@ -0,0 +1,11 @@
+./run.sh
+rm -f ../static/asciinema/demo.cast
+mv ./demo.cast ../static/asciinema/demo.cast
+```
````
does asciienma work with netlify? Im thinking maybe the Makefile being used by netlify will need an update?

michaelawyu replied:
For this part, since we probably do not need to update the asciinema cast very often, this script is not executed at the time of deployment. A finished cast is already checked in with this PR, which will be uploaded as a static file :)

bobcatfish commented on sync/pipelines.yaml:
```diff
@@ -0,0 +1,68 @@
+#  - pipelines.md: pipelines.md
+#  - resources.md: resources.md
+#  - taskruns.md: taskruns.md
+#  - tasks.md: tasks.md
```
do we want to include the commented out sections?

michaelawyu replied:
I included these commended out sections as an example for showcasing how to set up multiple document versions, maybe they are better presented as explanatory texts; I will update them.

michaelawyu commented:
> Looked over all the files and code looks good. Only comment I have is that I saw some duplicate inline styling in the html files as well as adding them via js DOM methods. I feel these styles should go separate in their respective css files so in the future it would be easier to edit/update the styling. All in all, very good work Chen :)

Hi @carlos-logro ,

Thanks for the review :) I'll fix issue 1-4 as soon as possible. Unfortunately for issue 5 and 6, the Katacoda embed is not built by me and hosted on the Katacoda site; I'll dig around for a proper solution but as far as I know the script does not offer many customization options.

michaelawyu

michaelawyu commented:
Hi @bobcatfish ,

Ideally Netlify will run the helper/sync script at the time of deployment via the Makefile, which syncs contents and prepares a few scripts, the latter of which is required by the Hugo framework for pre-processing. A part of script downloads the background image from a Google Cloud Storage bucket, which requires GCP credentials; I have to put the image there as the licensing term requires that we do not check the image in a publicly accessible location (e.g. a GitHub repo). Since Netlify is not set up with the credential, the script will not run and no file is generated, hence the error reported. 

To fix this, I was wondering if I can have (temporary) access to the Netlify instance. I vaguely recall that you have granted me access a while ago but at this moment when I signed into Netlify with my GitHub account I see no available projects :(

michaelawyu

vdemeester commented:
/hold

sbwsg commented on assets/js/tabs.js:
```diff
@@ -0,0 +1,24 @@
+$('.tab-content').find('.tab-pane').each(function (idx, item) {
```
curious why we're mixing jquery dom traversal with browser APIs in e.g. try.js.  Is there a reason not to consistently use one or other?

michaelawyu replied:
I have refactored the scripts. The style differs originally because I reused some scripts from an earlier project.

sbwsg commented on assets/js/tabs.js:
```diff
@@ -0,0 +1,24 @@
+
+$('.code-tabs ul.nav-tabs').each(function () {
+  $(this).find("li:first").addClass('active');
+})
```
v minor nit: missing semicolon

michaelawyu replied:
Done. Thanks :)

sbwsg commented on layouts/partials/navbar-vault.html:
```diff
@@ -0,0 +1,32 @@
+				{{ $url := urls.Parse .URL }}
+                {{ $baseurl := urls.Parse $.Site.Params.Baseurl }}
+                {{ if and (ne .Name "Documentation") (ne .Name "Try") }}
+                <a class="nav-link{{if $active }} active{{end}}" href="{{ with .Page }}{{ .RelPermalink }}{{ else }}{{ .URL | relLangURL }}{{ end }}" {{ if ne $url.Host $baseurl.Host }}target="_blank" {{ end }}><span{{if $active }} class="active"{{end}}>{{ .Name }}</span></a>
```
We're adding the "active" class to both the a and the span here if `$active` is truthy.  Would it be enough to simply have a CSS selector like:

```css
a.active { /* active a styles */ }
a.active > span { /* span inside active a styles */ }
```

Then the template logic here would be a tad easier to read.

michaelawyu replied:
For this part I actually reused the logic in the base theme. I totally see your point though.

sbwsg commented on layouts/partials/sidebar-vault.html:
```diff
@@ -0,0 +1,14 @@
+{{ $mid := printf "m-%s" (.RelPermalink | anchorize) }}
+<script>
+ $(function() {
+ $("#td-sidebar-menu #{{ $mid }}").toggleClass("active"); 
```
minor nit: indenting is a bit off here.  Suggest four spaces (or a tab if preferred) on this line and next.

michaelawyu replied:
Done. Thanks!

sbwsg commented on templates/_index.md.template:
```diff
@@ -0,0 +1,15 @@
+linkTitle: "Vault"
+---
+
+This section keeps the documentation of past versions of Tekton components.
```
minor nit: suggest "provides" instead of "keeps".

michaelawyu replied:
Done :) Thanks!

bobcatfish commented:
update: spoke with @michaelawyu and I think we've settled the netlify access he needs! @michaelawyu let me know if i can help in any other way

michaelawyu commented on assets/js/playground.js:
```diff
@@ -0,0 +1,53 @@
+  document.getElementById('katacoda-button').removeAttribute('href');
+  document.getElementById('github-button').removeAttribute('href');
+  document.getElementById('qwiklabs-button').removeAttribute('href');
+}
```
The scripts have been refactored.

michaelawyu commented on assets/js/try.js:
```diff
@@ -0,0 +1,46 @@
+  const katacodaSrc = article.getAttribute('data-katacoda-src');
+  const katacodaLnk = getKatacodaLnk(katacodaSrc);
+  const githubLnk = article.getAttribute('data-github-src');
+  const qwiklabsLnk = article.getAttribute('data-qwiklabs-src');
```
It should display correct in the new preview (the Netlify one) now. The QwikLabs version is not yet available, they need to go through a separate review process. I have grayed them out.

michaelawyu commented on assets/js/playground.js:
```diff
@@ -0,0 +1,53 @@
+function getKatacodaLnk (katacodaSrc) {
+  const elems = katacodaSrc.split('/');
+  return `${elems[0]}/scenarios/${elems[1]}`;
```
Done. Thanks!

michaelawyu commented on layouts/shortcodes/blocks/showcase.html:
```diff
@@ -0,0 +1,29 @@
+			<div class="row pl-0">
+				<div class="col-lg-6 mb-5 mb-lg-0">
+					<div class="text-left">
+						<p class="h3 text-light d-inline">Tekton is a powerful and flexible </p>
```
I am using 929292 now for the light color variable.

michaelawyu commented:
Netlify instance is now configured. Preview should be working now.

Added:
- Netlify support
- Cookie notice
- Design tweaks for the homepage
- Try page refactored (using the same CSS framework as other pages now)
- New partners section (no logos)
- Updated sync scripts
- Refactored Javascripts

and other minor fixes.

bobcatfish commented on layouts/shortcodes/blocks/partners-alternate.html:
```diff
@@ -0,0 +1,27 @@
+                    <a href="https://www.ibm.com/">IBM</a>,
+                    <a href="https://puppet.com/">Puppet</a>,
+                    <a href="https://www.redhat.com/">Red Hat</a>,
+                    and <a href="https://triggermesh.com/">Trigger Mesh</a>.</h3>
```
i like using the friends logo!! :D

some of these folks dont contribute to Tekton, but they are in tekton/friends as users - it would maybe be more safe to say something like "Tekton is created by the CDF" and leave it at that

michaelawyu replied:
Done. I have rephrased it as "Tekton is a collaborative project where members of the ecosystem contribute together to make CI/CD easier for everyone. See the individuals and organizations that are involved in the Tekton project."

bobcatfish replied:
Great!!! :D

anneqm commented on content/en/_index.html:
```diff
@@ -1,16 +1,67 @@
-{{% /blocks/lead %}}
+{{< blocks/section color="danger" >}}
+{{% blocks/feature icon="fas fa-globe-asia" title="Standardization" %}}
+Tekton standardizes CI/CD tooling and processes accorss vendors, languages,
```
Minor typo. Please correct spelling `accorss` should be `across`

michaelawyu replied:
Done. Thanks!

anneqm commented on content/en/_index.html:
```diff
@@ -1,16 +1,67 @@
+
+
+{{% blocks/feature icon="fas fa-medal" title="Built-in best practices" %}}
+Tekton let you create cloud-native CI/CD systems quickly, leveraging advanced
```
Minor typo `Tekton let you` should be `Tekton lets you`... OR consider re-using the wording from the similar sections on the documentation page.

michaelawyu replied:
Done. Thanks!

bobcatfish commented:
I think that some of the most recent changes might not have been pushed @michaelawyu since the preview still says "Tekton is created by Google in partnership with Alibaba, Ant Financial, CloudBees, Elastic, IBM, Puppet, Red Hat, and Trigger Mesh."

I think we're ready to merge once we push those last few changes tho! :pray:

caniszczyk commented:
cool tool, I would definitely reword that, also make note of the CDF website guidelines:
https://github.com/cdfoundation/foundation/blob/master/project-website-guidelines.md

I would say "Tekton was originally created by Google and contributed to the CDF in partnership with Alibaba, Ant Financial, CloudBees, Elastic, IBM, Puppet, Red Hat, and Trigger Mesh."

You can use some of the wording/design from Envoy as inspiration potentially: https://www.envoyproxy.io/

michaelawyu commented:
@bobcatfish All done ;)
@caniszczyk Thanks! I was in a discussion about it the other day; it seemed that the concern is, some of the partners listed in friends are not code contributors to the project, and the current expression might be a little confusing in that aspect. I have used a more neutral expression, PTAL.

bobcatfish commented:
Thanks @michaelawyu !! Let's do this :D

/approve
/lgtm

![image](https://user-images.githubusercontent.com/432502/74660239-c19b1600-5163-11ea-87ef-bada78199d1b.png)

bobcatfish commented:
/hold cancel

popcor255 commented on layouts/partials/navbar-version-selector-docs.html:
```diff
@@ -0,0 +1,19 @@
+{{ $jsVerSwitcher := resources.Get "js/version-switcher.js" }}
```
This file is not in this repo. Is it missing or just somewhere else? If I remove it I can run it dev, but I do not want to make pr removing it without knowing what it does.

michaelawyu replied:
This file is dynamically generated at the time of deployment. To run locally, use the sync script in the repository.

popcor255 replied:
thank you
